core/types/block.libevm.go

```diff
@@ -108,3 +109,93 @@ func (*NOOPHeaderHooks) DecodeRLP(h *Header, s *rlp.Stream) error {
 	type withoutMethods Header
 	return s.Decode((*withoutMethods)(h))
 }
+
+var (
+	_ interface {
+		rlp.Encoder
+		rlp.Decoder
+	} = (*Body)(nil)
+
+	// The implementations of [Body.EncodeRLP] and [Body.DecodeRLP] make
+	// assumptions about the struct fields and their order, which we lock in here as a change
+	// detector. If this breaks then it MUST be updated and the RLP methods
+	// reviewed + new backwards-compatibility tests added.
+	_ = &Body{[]*Transaction{}, []*Header{}, []*Withdrawal{}}
+)
+
+// EncodeRLP implements the [rlp.Encoder] interface.
+func (b *Body) EncodeRLP(dst io.Writer) error {
+	w := rlp.NewEncoderBuffer(dst)
+
+	return w.InList(func() error {
+		if err := rlp.EncodeListToBuffer(w, b.Transactions); err != nil {
+			return err
+		}
+		if err := rlp.EncodeListToBuffer(w, b.Uncles); err != nil {
+			return err
+		}
+
+		hasLaterOptionalField := b.Withdrawals != nil
+		if err := b.hooks().AppendRLPFields(w, hasLaterOptionalField); err != nil {
```

Comment:
What's the use for the second argument `hasLaterOptionalField`?

Reply:
https://pkg.go.dev/github.com/ethereum/go-ethereum/rlp#:~:text=When%20encoding%20a%20struct%20with%20optional%20fields%2C%20the%20output%20RLP%20list%20contains%20all%20values%20up%20to%20the%20last%20non%2Dzero%20optional%20field.

It's too allow for decoding such that the nth RLP item is always the nth struct field. You'll see in the hook that the parameter is called `mustWriteEmptyOptional`. It's not necessary in `coreth` because there aren't any optional fields in the extras, but once we're exposing this I think it's good software citizenship to not set a foot trap for some unsuspecting third-party user.